### PR TITLE
doc: Add separate productivity notes document

### PR DESCRIPTION
**Pull request by dongcarl**
Many developers have their own tools and tricks to be more productive
during their cycles, so let's document the best ones so that everyone
can benefit from them.

### CODE REVIEW

File: doc/productivity.md [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+General
+------
+
+### Cache compilations with `ccache`
+
+The easiest way to faster compile times is to cache compiles. `ccache` is a way to do so, from its description at the time of writing:
+
+> ccache is a compiler cache. It speeds up recompilation by caching the result of previous compilations and detecting when the same compilation is being done again. Supported languages are C, C++, Objective-C and Objective-C++.
+
+Install `ccache` through your distribution's package manager, and follow the symlinks method [here](https://ccache.samba.org/manual/latest.html#_run_modes) to set it up.
```
Comment (MarcoFalke) [inline doc/productivity.md]: I don't think it is required to do any fancy symlinking. Our configure will pick up ccache on its own.

File: doc/productivity.md [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+If you have multiple threads on your machine, you can tell `make` to utilize all of them with:
+
+```sh
+make -j$(nproc)
+```
+
+If you want `make` to not limit the number of jobs that can run simultaneously, do:
+
+```sh
+make -j
```
Comment (MarcoFalke) [inline doc/productivity.md]: That could easily OOM your machine ;)

File: doc/productivity.md [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+Reviewing code
+--------------
+
+### Reduce mental load with `git diff` options
+
+When reviewing patches which change indentation in C++ files, use `git diff -w` and `git show -w`. This makes the diff algorithm ignore whitespace changes. This feature is also available on github.com, by adding `?w=1` at the end of any URL which shows a diff.
+
+When reviewing patches that change symbol names in many places, use `git diff --word-diff`. This will instead of showing the patch as deleted/added *lines*, show deleted/added *words*.
+
+When reviewing patches that move code around, try using `git diff --patience commit~:old/file.cpp commit:new/file/name.cpp`, and ignoring everything except the moved body of code which should show up as neither `+` or `-` lines. In case it was not a pure move, this may even work when combined with the `-w` or `--word-diff` options described above.
```
Comment (MarcoFalke) [inline doc/productivity.md]: for moved code, there'd also be `--color-moved=dimmed-zebra`

**Review by maflcko — APPROVED** [review]

File: doc/productivity.md (hunk 3 above)
Comment (MarcoFalke) [reply doc/productivity.md]: example:

```
git diff 2ca632e5b44a8385989c8539cc4e30e60fdee16c~ 2ca632e5b44a8385989c8539cc4e30e60fdee16c --color-moved=dimmed-zebra src/test

**Comment by laanwj** [comment]
Concept ACK, though sometimes I wonder if a wiki wouldn't be a better place for things like this. It's not that easy to maintain documentation in the repository as it has to go through the same review bottleneck as everything else.

But despite that, thanks for writing this!

**Comment by dongcarl** [comment]
@laanwj I believe GitHub allows you to create Wikis for repos, we could try that in the future :relaxed:

**Comment by hebasto** [comment]
Concept ACK.

File: doc/productivity.md (hunk 2 above)
Comment (laanwj) [reply doc/productivity.md]: Unlimited is very likely to cause OOM problems with C++. Isn't it usually the recommendation to choose number_of_cores+1 as parallelism ? (unless for low-memory/embedded machines)

**Comment by dongcarl** [comment]
Addressed comments.

File: doc/productivity.md [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+General
+------
+
+### Cache compilations with `ccache`
+
+The easiest way to faster compile times is to cache compiles. `ccache` is a way to do so, from its description at the time of writing:
+
+> ccache is a compiler cache. It speeds up recompilation by caching the result of previous compilations and detecting when the same compilation is being done again. Supported languages are C, C++, Objective-C and Objective-C++.
+
+Install `ccache` through your distribution's package manager, and run `./configure` with your normal flags to pick it up.
```
Comment (MarcoFalke) [inline doc/productivity.md]:
```
this diff appears to have added new lines with trailing whitespace.
The following changes were suspected:
diff --git a/doc/productivity.md b/doc/productivity.md
@@ -0,0 +1,161 @@
+
+Install `ccache` through your distribution's package manager, and run `./configure` with your normal flags to pick it up.
^---- failure generated from test/lint/lint-whitespace.sh

**Comment by MeshCollider** [comment]
Concept ACK, looks good so far

**Pull request merged**